lib/Dialect/Handshake/HandshakeOps.cpp

```diff
@@ -168,12 +169,48 @@ struct EliminateUnusedForkResultsPattern : mlir::OpRewritePattern<ForkOp> {
     return success();
   }
 };
+
+struct EliminateForkToForkPattern : mlir::OpRewritePattern<ForkOp> {
+  using mlir::OpRewritePattern<ForkOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ForkOp op,
+                                PatternRewriter &rewriter) const override {
+    auto parentForkOp =
+        dyn_cast_or_null<ForkOp>(op.getOperand().getDefiningOp());
```

Comment:
Nit: I think you can say getDefiningOp<ForkOp> and it will use dyn_cast_or_null under the hood.